### ISSUE TITLE
Add product image selection struct

### ISSUE THREAD

**Issue by vsupalov** [stackabletech/operator-rs#470]
Implementation of design described in [ADR023](https://docs.stackable.tech/home/stable/contributor/adr/ADR023-product-image-selection.html). Create reusable code, which will be integrated into all operators to enable new product/operator version specification.

This task is related to two epics:

* https://github.com/stackabletech/issues/issues/256
* https://github.com/stackabletech/issues/issues/257

The first epic is the primary focus at the moment, as it is a major blocker to get to multi-arch.

The feature will be implemented to be optional for now.

## Acceptance criteria

* [ ] Capture requirements described in [ADR](https://docs.stackable.tech/home/stable/contributor/adr/ADR023-product-image-selection.html) as test cases for the feature inside of operator-rs
  * All use cases and config examples described in "Option 6"
  * All examples described in "Decision Drivers"
* [ ] Operator using this in the future, will tell the function what product it operates on (to build the image name)
* [ ] Provide helper functions in operator-rs to replace existing helper functions like image_version and product_version in operator codebases
* [ ] Implement `pullPolicy` as well
* [ ] Implement `stackableVersion` case handling, as described in ADR Option 6
* [ ] Implement `custom` case handling, as described in ADR Option 6
  * limit functionality to public repositories
  * OPTIONAL: Implement `pullSecrets` aspects of ADR, and using private repositories
* [ ] Implement erroring when user tries to use `stackable` format for now
* [ ] Use changes in https://github.com/stackabletech/kafka-operator/issues/477 to verify that the solution is viable and usable

## Related out-of-scope followup tasks

To be performed in https://github.com/stackabletech/issues/issues/257

* If not implemented above: `pullSecrets` aspects of ADR, and using private repositories

# Status
* Needs work in kube-rs => https://github.com/kube-rs/kube/pull/1028
* Needs bump of kube-rs to 0.75.0 (and higher later on) in operator-rs to use the new feature => https://github.com/stackabletech/operator-rs/issues/471
* PR => https://github.com/stackabletech/operator-rs/pull/476
* PR for reference implementation in kafka-operator => https://github.com/stackabletech/kafka-operator/pull/482

**Comment by lfrancke** [comment]
Is this an epic or a task?

**Comment by vsupalov** [comment]
This is a task, which will be used in https://github.com/stackabletech/kafka-operator/issues/477 and an upcoming epic "Apply changes which were found out to make Kafka operator stackable-experimental friendly to all operators" for https://github.com/stackabletech/issues/issues/256

**Comment by sbernauer** [comment]
Implementation is basically done. Draft PR for operator-rs and for kafka-operator are linked in ticket description.
We are blocked on the following things (also linked in ticket description):
* kube-rs PR to add support for untagged enums
* operators-rs PR to bump kube-rs version

**Comment by sbernauer** [comment]
Currently waiting for kube-rs 0.76.0
New operator-rs version has the fragment validation PR in. The Kafka spike needs to be updated somehow

**Comment by sbernauer** [comment]
Everything should be done now, moving to "Waiting for Review"

**Comment by sbernauer** [comment]
Implemented via https://github.com/stackabletech/operator-rs/commit/863d53ec5f8fb883b80e59c934310f9c296a33e5

**Comment by lfrancke** [comment]
None of the acceptance criteria have been checked.
Could you double check that everything has been done?

**Comment by sbernauer** [comment]
checked
